woocommerce/woocommerce-gateway-stripe
# Fix Uncaught TypeError: Argument 2 passed to WC_Stripe_API::detach_payment_method_from_customer() errors

Fixes #2615 

## Changes proposed in this Pull Request:

We've seen a few reports of merchants experiencing the following errors:

```
PHP Notice:  Undefined property: stdClass::$source in /plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php on line 137
PHP Notice:  Trying to get property 'id' of non-object in /plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php on line 137
PHP Fatal error:  Uncaught TypeError: Argument 2 passed to WC_Stripe_API::detach_payment_method_from_customer() must be of the type string, null given, called in /plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-customer.php on line 472 and defined in /home/site/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-api.php:327
```

The errors occurred because the customer would fail the 3D-Secure challenge and we'd try to make sure that the payment method/source was detached from the customer in Stripe, however we were pulling the payment method ID from the intent error in a way that is not compatible with newer payment intent errors. ie we were accessing the `$intent->last_payment_error->source->id` rather than the `$intent->last_payment_error->payment_method->id`.

This PR fixes that by ensuring we fetch the Payment method iD from the intent correctly, but also ensures we don't cause `Uncaught TypeError: Argument 2 passed to WC_Stripe_API::detach_payment_method_from_customer()` errors.

## Testing instructions

1. Enable the **legacy checkout experience** in your Stripe settings.
2. Place a product in your cart. 
3. On the checkout enter a 3Ds card like `4000 0000 0000 3063`
4. Check the box to save the payment method. 
5. Place the order.
6. Wait until the 3D-Secure modal pops up.
7. Fail the 3D-Secure challenge. 
   - On `develop` Check your php error logs and you should see the errors I mentioned above. 
   - On this branch there should be no errors. 

---


**Post merge**
